### ISSUE TITLE
Client disconnect mid-stream + subsequent model load crashes entire server (libc++abi: no Stream(gpu,N) in current thread)

### ISSUE THREAD

**Issue by dpalmqvist** [motsognirr/olmlx#651]
## Summary

A client disconnecting mid-stream during `/api/chat` (or any streaming
generation), followed by **any** subsequent request that triggers a model
load/eviction (e.g. `/v1/audio/transcriptions` with a Whisper model),
crashes the entire server process with an uncaught C++ exception:

```
libc++abi: terminating due to uncaught exception of type std::runtime_error: There is no Stream(gpu, 6) in current thread.
```

This is a hard process abort (`libc++abi`), not a Python exception — it
cannot be caught by any exception handler and takes down the whole
`uvicorn` process, dropping all in-flight and future requests. A client
closing a streaming connection early (e.g. a browser tab close, a CLI
`Ctrl-C`, a proxy timeout) is a completely ordinary occurrence, so this is
trivially triggerable in normal use, not just adversarial testing.

## Reproduction

Server: `uv run olmlx serve` (localhost:11434), models already pulled:
`mlx-community/Qwen2.5-0.5B-Instruct-4bit`, `mlx-community/whisper-large-v3-turbo`.

1. Start a streaming chat request and kill the client mid-generation
   (before it finishes):
   ```bash
   ( curl -s http://localhost:11434/api/chat -d '{
       "model": "mlx-community/Qwen2.5-0.5B-Instruct-4bit",
       "messages": [{"role":"user","content":"Write a very long essay about the history of computing, at least 1000 words."}],
       "stream": true
     }' > /tmp/out.jsonl ) &
   CURL_PID=$!
   sleep 0.4
   kill -9 $CURL_PID
   wait
   ```
   Server log shows the (handled) deferred-cleanup path:
   ```
   [WARNING] olmlx.engine.inference: Inference thread still alive after cleanup attempts — deferring Metal sync and lock release until thread exits
   [INFO] olmlx.engine.inference: Deferred inference cleanup: thread exited cleanly
   [INFO] olmlx.engine.inference: Deferred inference cleanup: lock released
   ```

2. Immediately after (~1s later), send any request that loads a
   different model, e.g.:
   ```bash
   curl -s http://localhost:11434/v1/audio/transcriptions \
     -F "file=@/tmp/test_audio.wav" \
     -F "model=mlx-community/whisper-large-v3-turbo"
   ```

## Observed

The server process dies immediately after evicting the previous model and
loading Whisper — no HTTP response is returned (`curl`: "empty reply from
server" / connection refused on the next request), and the process is
gone (`pgrep -fl olmlx` → nothing).

Server log tail:
```
2026-07-11 08:25:33 [INFO] olmlx.engine.inference: [8c48c825] Prompt cache init: creating fresh cache for 49 tokens
INFO:     127.0.0.1:63118 - "POST /api/chat HTTP/1.1" 200 OK
2026-07-11 08:25:33 [WARNING] olmlx.engine.inference: [8c48c825] Inference thread still alive after cleanup attempts — deferring Metal sync and lock release until thread exits
2026-07-11 08:25:33 [INFO] olmlx.engine.inference: [8c48c825] Deferred inference cleanup: thread exited cleanly
2026-07-11 08:25:33 [INFO] olmlx.engine.inference: [8c48c825] Deferred inference cleanup: lock released
2026-07-11 08:25:35 [INFO] olmlx.engine.model_manager: [e5d73233] Normalized model name 'mlx-community/whisper-large-v3-turbo' -> 'mlx-community/whisper-large-v3-turbo:latest'
2026-07-11 08:25:35 [INFO] olmlx.engine.model_manager: [e5d73233] Evicting model mlx-community/Qwen2.5-0.5B-Instruct-4bit:latest
2026-07-11 08:25:35 [INFO] olmlx.engine.model_manager: [e5d73233] Loading model mlx-community/whisper-large-v3-turbo:latest from mlx-community/whisper-large-v3-turbo
2026-07-11 08:25:35 [INFO] olmlx.engine.model_manager: [e5d73233] Detected model kind for mlx-community/whisper-large-v3-turbo: whisper
2026-07-11 08:25:35 [INFO] olmlx.engine.model_manager: [e5d73233] Model mlx-community/whisper-large-v3-turbo:latest caps: tools=False, thinking=False, thinking_tags=False
libc++abi: terminating due to uncaught exception of type std::runtime_error: There is no Stream(gpu, 6) in current thread.
```

Reproduced twice independently (different GPU stream indices — `gpu, 0`
and `gpu, 6` — across runs), and once via a slightly different path
(killed stream → failed Kokoro TTS load → Whisper transcription → same
crash signature), so the trigger is the leftover state from the
mid-stream disconnect, not the specific downstream model.

## Expected

The server should survive a client disconnecting mid-stream. Whatever
per-thread Metal stream state the deferred-cleanup path leaves behind
(per `CLAUDE.md`'s documented thread-local `generation_stream` /
`default_stream` invariants, and the mlx 0.32.0 thread-local streams
change from #499) should not leak into or corrupt the stream context that
a subsequent, unrelated model-load/generation thread expects to find —
or if the deferred-cleanup thread genuinely needs to fully finish before
it's safe to load another model, the model manager should wait for it
rather than racing ahead into a bad stream state that aborts the process.

## Environment

- Models: `mlx-community/Qwen2.5-0.5B-Instruct-4bit` (chat),
  `mlx-community/whisper-large-v3-turbo` (transcription)
- `uv run olmlx serve` on macOS / Apple Silicon
- Not the documented SIGBUS-at-teardown non-bug (#470) — this is a crash
  *during* live request handling, with a distinct `std::runtime_error`
  signature, not a teardown-time SIGBUS.

**Comment by dpalmqvist** [comment]
Reproduced the same crash signature independently, and found the trigger is broader than a mid-stream disconnect.

## Repro without any client disconnect

After a normal exploratory sweep of ~40 requests across `/api/chat`, `/v1/chat/completions`, `/v1/messages`, `/v1/responses`, tool-calling, JSON-schema grammar, and rerank (no killed/aborted streams anywhere), the very next two requests were:

1. `/v1/audio/speech` with `prince-canuma/Kokoro-82M` — failed fast with a Python `ImportError` (misaki not installed) *before* any generation/Metal work happened, just a model load + immediate pipeline-init exception.
2. `/v1/audio/transcriptions` with `mlx-community/whisper-large-v3-turbo` — crashed the whole process:

```
2026-07-15 08:33:49 [INFO] olmlx.engine.model_manager: [0d995029] Evicting model prince-canuma/Kokoro-82M:latest
2026-07-15 08:33:50 [INFO] olmlx.engine.model_manager: [0d995029] Loading model mlx-community/whisper-large-v3-turbo:latest from mlx-community/whisper-large-v3-turbo
2026-07-15 08:33:50 [INFO] olmlx.engine.model_manager: [0d995029] Detected model kind for mlx-community/whisper-large-v3-turbo: whisper
2026-07-15 08:33:50 [INFO] olmlx.engine.model_manager: [0d995029] Model mlx-community/whisper-large-v3-turbo:latest caps: tools=False, thinking=False, thinking_tags=False
libc++abi: terminating due to uncaught exception of type std::runtime_error: There is no Stream(gpu, 0) in current thread.
```

`ps` confirmed the server process was gone afterward — same hard abort as reported here.

I then tried to reproduce this exact same 2-request sequence (Kokoro-fails → Whisper-transcribe) on a **freshly started** server (first and second request of the session) three separate times, and it succeeded cleanly every time — no crash. So the disconnect isn't the load-bearing trigger; what differs is how many prior requests have driven `asyncio.to_thread` calls through the default executor's thread pool. On a fresh server, the whisper *load* thread and the whisper *transcribe* thread are very likely the same pool thread (nothing else has churned the pool), so the crash is masked. After enough prior traffic has rotated more threads into the pool, load and transcribe land on different OS threads and the crash reproduces — with no disconnect required.

## Root cause candidate

This looks like the same underscore-buffer gap already fixed for scaled-RoPE (`_freqs`) and documented in CLAUDE.md, but not yet applied to the Whisper loader chokepoint (`ModelManager._load_with_model_type_fallback`, the `kind == "whisper"` branch at `olmlx/engine/model_manager.py:3586-3594`), which calls `mlx_whisper.load_models.load_model()` directly and returns without running `_materialize_module_buffers`.

`mlx_whisper.load_models.load_model()` does call `mx.eval(model.parameters())`, but `nn.Module.parameters()` skips underscore-prefixed attributes (per the documented gap), and `mlx_whisper.whisper.Whisper`'s submodules stash two lazily-computed buffers under underscore keys:

- `AudioEncoder.__init__`: `self._positional_embedding = sinusoids(n_ctx, n_state).astype(dtype)` — built from `mx.exp`/`mx.arange`/`mx.sin`/`mx.cos`/`mx.concatenate`, never eval'd.
- `TextDecoder.__init__`: `self._mask = nn.MultiHeadAttention.create_additive_causal_mask(n_ctx).astype(dtype)` — same gap.

Both stay lazy and bound to whichever thread first evaluates a graph referencing them (the load thread, since nothing else touches them at load time). When the *transcribe* worker thread (a separate `asyncio.to_thread` call in `generate_transcription`, `olmlx/engine/inference.py:5101`) runs the encoder/decoder forward pass on a different pool thread, evaluating those two buffers raises `There is no Stream(gpu, N) in current thread` — exactly the observed signature, and exactly the same class of bug as the already-fixed scaled-RoPE case, just on the whisper load path instead of the mlx-lm text path.

If this diagnosis is right, the fix would be to run the existing `_materialize_module_buffers(model)` helper (or equivalent) on the Whisper model right after `whisper_loader.load_model()` in `model_manager.py`, the same way it's wired at the mlx-lm/mlx-vlm/flash chokepoints.

Environment: same as above (macOS/Apple Silicon, `uv run olmlx serve`, `mlx-community/whisper-large-v3-turbo`).
